managementAgent.js

```diff
@@ -1,51 +1,130 @@
-const WebSocket = require('ws');
 const Uuid = require('uuid');
+const WebSocket = require('ws');
+const metadata = require('./lib/metadata/wrapper');
 
+const _config = require('./lib/Config').config;
 const logger = require('./lib/utilities/logger');
 const { initManagement } = require('./lib/management');
-const _config = require('./lib/Config').config;
+const { managementAgentMessageType } = require('./lib/management/agentClient');
+const { addOverlayMessageListener } = require('./lib/management/push');
+
+
+//TODO: auth?
+
+const WAIT_INIT_DURATION_MS = 5000;
+const CHECK_BROKEN_CONNECTIONS_FREQUENCY_MS = 15000;
+
+
+class ManagementAgentServer {
+    constructor() {
+        this.port = _config.managementAgent.port || 8010;
+        this.wss = null;
+        this.loadedOverlay = null;
+
+        process.on('SIGINT', this.stop.bind(this));
+        process.on('SIGHUP', this.stop.bind(this));
+        process.on('SIGQUIT', this.stop.bind(this));
+        process.on('SIGTERM', this.stop.bind(this));
+        process.on('SIGPIPE', () => {});
+    }
+
+    start() {
+        /* Define REPORT_TOKEN env variable needed by the management
+         * module. */
+        process.env.REPORT_TOKEN = process.env.REPORT_TOKEN
+          || _config.reportToken
+          || Uuid.v4();
+
+        metadata.setup(error => {
+            if (error) {
+                const msg = `failed to setup metadata: ${error}`;
+                throw new Error(msg);
```

Comment:
We should not throw in async callbacks, instead call a callback with the failure so that the toplevel code can schedule a retry